node/src/reactor/main_reactor/tests.rs

```diff
@@ -309,90 +349,120 @@ async fn run_equivocator_network() {
-        .find(|reactor| *reactor.inner().consensus().public_key() == alice_public_key)
-        .unwrap()
-        .set_filter(move |event| {
-            let now = Timestamp::now();
-            match &event {
-                MainEvent::ConsensusMessageIncoming { .. } => {}
-                MainEvent::NetworkRequest(
-                    NetworkRequest::SendMessage { payload, .. }
-                    | NetworkRequest::ValidatorBroadcast { payload, .. }
-                    | NetworkRequest::Gossip { payload, .. },
-                ) if matches!(**payload, Message::Consensus(_)) => {}
-                _ => return Either::Right(event),
-            };
-            let first_message_time = *maybe_first_message_time.get_or_insert(now);
-            if now < first_message_time + min_round_len * 3 {
-                return Either::Left(time::sleep(min_round_len.into()).event(move |_| event));
-            }
-            Either::Right(event)
-        });
+
+    let mut alice_reactors = net
+        .reactors_mut()
+        .filter(|reactor| *reactor.inner().consensus().public_key() == alice_public_key);
+
+    // Delay all messages to and from the first of Alice's nodes until three rounds after the first
+    // message.  Further, significantly delay any incoming pings to avoid the node detecting the
+    // doppelganger and deactivating itself.
+    alice_reactors.next().unwrap().set_filter(move |event| {
+        if is_ping(&event) {
+            return Either::Left(time::sleep((min_round_len * 30).into()).event(move |_| event));
+        }
+        let now = Timestamp::now();
+        match &event {
+            MainEvent::ConsensusMessageIncoming(_) => {}
+            MainEvent::NetworkRequest(
+                NetworkRequest::SendMessage { payload, .. }
+                | NetworkRequest::ValidatorBroadcast { payload, .. }
+                | NetworkRequest::Gossip { payload, .. },
+            ) if matches!(**payload, Message::Consensus(_)) => {}
+            _ => return Either::Right(event),
+        };
+        let first_message_time = *maybe_first_message_time.get_or_insert(now);
+        if now < first_message_time + min_round_len * 3 {
+            return Either::Left(time::sleep(min_round_len.into()).event(move |_| event));
+        }
+        Either::Right(event)
+    });
+
+    // Significantly delay all incoming pings to the second of Alice's nodes.
+    alice_reactors.next().unwrap().set_filter(move |event| {
+        if is_ping(&event) {
+            return Either::Left(time::sleep((min_round_len * 30).into()).event(move |_| event));
+        }
+        Either::Right(event)
+    });
+
+    drop(alice_reactors);
 
     let era_count = 4;
 
     let timeout = Duration::from_secs(90 * era_count);
     info!("Waiting for {} eras to end.", era_count);
-    net.settle_on(&mut rng, is_in_era(EraId::new(era_count)), timeout)
-        .await;
+    net.settle_on(
+        &mut rng,
+        has_completed_era(EraId::new(era_count - 1)),
```

Comment:
Is this `- 1` now required?

Reply:
Yes - previously we'd wait until the nodes' consensus components reported they'd started era 4, but now we wait until the nodes' storage components report they've stored the switch block from era 3.  The change was required as not all nodes are necessarily validators and I needed a way to have a similar condition.